### DISCUSSION TITLE
Variable de clase o singleton

### DISCUSSION THREAD

**Discussion by francorojo** [dds-jv/foro#174] [Consultas de objetos] [answered]
Buenas, tenia una duda, en el caso de por ejemplo tener una clase llamada Contador, que representa comportamientos que haria un contador y que se necesitan varias instancias diferentes, tiene dentro de otras, la responsabilidad de calcular impuestos y tiene por ejemplo el metodo precioFinal(BienComercial objetoCatalogo) que devuelve su precio en base a los impuestos actuales del pais.

_Suponiendo que el metodo hace lo siguiente:
return objetoCatalogo * (1 + porcentajeIva)_

_Mi duda es: Cual seria una buena forma de solucion para almacenar el valor de porcentajeIva?_

**Se me ocurren estas alternativas:**

- Guardar porcentajeIva como una variable de clase y se pueda settear por parametro (Perooo esto implicaria que si cambia, hay q cambiarlo en cada instancia que hayamos creado, porque todos los contadores deben seguir el mismo valor)

- Guardar el porcentajeIva dentro de un singleton que conozca todos los impuestos del pais, el cual se inyecta por dependencia y en ese caso cambia para todos de forma instantanea cuando sea el momento (Pero no se si es mucha complejidad para solamente, en este caso, cambiar un valor por ahora)

- O guardar el valor porcentajeIva como variable de clase, y que cada instancia la use, entonces en este caso cuando tenga q cambiar se puede cambiar directamente en la variable de clase y cambia para todos (Pero no estoy seguro, es correcto esto? o hay un acoplamiento implicito?, por lo que lei en internet seria util desde el punto de vista que solo se crea en un espacio de memoria y no se crea para cada instancia)

- Tal vez puede ser una constante global del codigo (Desconozco esto bien como funciona, pero en esta solucion creo q habria q recompilar el codigo)

Muchas gracias si alguno me da una idea de cual puede ser mas util que las demas o si hay alguna mejor que esas q se me ocurrieron, saludoos!!!

**Comment by gastonprieto** [comment]
Buenas @francorojo!

Sin el enunciado es un poco dificil opinar, pero te comento lo que veo en base a lo que dijiste 👍 

> clase llamada Contador, que representa comportamientos que haria un contador

Esto me suena raro, porque un contador es un rol, parece como si estuvieras modelando al usuario
Ademas esto `precioFinal(BienComercial objetoCatalogo)` me suena como que tal vez debería ser: `objetoCatalogo.precioFinal()`  y msa si la implementación es así:

```Java
return objetoCatalogo * (1 + porcentajeIva)
```

Me huele raro (code smell)... pero nada con el contexto completo tal vez sirve que sea así, pero quiero dejar mi warning ⚠️ 

----

> Guardar porcentajeIva como una variable de clase y se pueda settear por parametro (Perooo esto implicaria que si cambia, hay q cambiarlo en cada instancia que hayamos creado, porque todos los contadores deben seguir el mismo valor)

Ojo, eso no es una variable de clase, eso es una variable de instancia. Una variable de clase es un atributo de la clase, no de la instancia y es global a todas las instancias, estos atributos de clase en Java se usan con `static`

En el caso de que hablaras de variable de instancia del contador, si deberías cambiarlo en todos los que corresponda, tal vez si tenes un repo de contadores puedas hacerlo, pero definitivamente es una desventaja. Como ventaja es que si queres tener contadores con distintos impuestos podes hacerlo.

>Guardar el porcentajeIva dentro de un singleton que conozca todos los impuestos del pais, el cual se inyecta por dependencia y en ese caso cambia para todos de forma instantanea cuando sea el momento (Pero no se si es mucha complejidad para solamente, en este caso, cambiar un valor por ahora)

Suena complejo, además deberías tener varios singletons, uno por cada país por lo que entiendo que decís.

> O guardar el valor porcentajeIva como variable de clase, y que cada instancia la use, entonces en este caso cuando tenga q cambiar se puede cambiar directamente en la variable de clase y cambia para todos (Pero no estoy seguro, es correcto esto? o hay un acoplamiento implicito?, por lo que lei en internet seria util desde el punto de vista que solo se crea en un espacio de memoria y no se crea para cada instancia)

Acá volves usar el termino variable de clase, en este caso de forma correcta 👍 . Es importante que esto lo uses de forma adecuada, porque las soluciones suelen implicar distintas cosas y después no se entiende 😕

Tratamos de no evaluar las decisiones en terminos de memoria, las optimizaciones quedan para otro momento. ℹ️ 

El tema que tenes con esto como te comentaba en el punto anterior, es que por lo que me decís es que contadores de distintos paises no podrían tener valores distintos de IVA. Entonces esta solución no cumple con los requerimientos.

> Tal vez puede ser una constante global del codigo (Desconozco esto bien como funciona, pero en esta solucion creo q habria q recompilar el codigo)

Una constante en Java esta modelada como un atributo de clase que no cambia. Así que es parecida a la de arriba. Si quisieras varios valores bien conocidos en tu código son los enums, y a cada enum asociarle un valor en particular (revisate el apunte de java para programadores objetosos que ahi explica como usar enums de forma un poco mas interesante).

---

Creo que hay una opción que no ponderaste. Me dijiste:

> precio en base a los impuestos actuales del pais

Y si modelas los objetos país? y a estos les asocias el IVA.

**Comment by francorojo** [comment, marked as answer]
Muchas graciaaas! Nono la verdad que me surgio la duda y quise poner un ejemplo para preguntar, no estaba haciendo un ejercicio en particular. (Ahora que me lo decis, es verdad que suena mucho mas logico tener la responsabilidad en el objeto del producto y en este caso modelar ademas el iva dentro del pais)
Pero me re sirve lo q me dijiste de cada opcion para comparar y tener q decidir depende el caso! Con la forma de la variable de instancia tengo q cambiarlas en todos lados o aprovechar que pueden ser diferentes, la variable de clase buenisimo q sea una opcion pero a tener cuidado, los singleton hay q tener mucho cuidado y ños enums son una herramienta re util al parecer, mas al poder otorgarle comportamiento, es como una constante con funcionalidad.
Saludooos!
